language/stdlib/modules/Libra.move

```diff
@@ -700,71 +700,159 @@ module Libra {
-        // If the AddCurrency privilege is on an address different from
-        // currency_addr(), the currency will appear not to be registered.
-        // If you change next to "true", prover will report an error.
-        pragma verify = false;
-
-        // SPEC: After register_currency, the currency is an official currency.
-        ensures spec_is_currency<CoinType>();
+    /// ## Management of capabilities
+
+    spec schema OnlyTCHasMintCapabilityInvariant {
+        /// Before a currency is registered, there is no mint capability for that currency.
+        invariant module forall coinType: type, addr: address:
+            !spec_is_currency<coinType>() ==> !exists<MintCapability<coinType>>(addr);
+
+        /// After a currency is registered, only the treasury compliance account
+        /// has the mint capability for that currency.
+        invariant module forall coinType: type, addr: address
+        where spec_is_currency<coinType>():
+            exists<MintCapability<coinType>>(addr)
+                ==> addr == spec_treasury_compliance_addr();
     }
 
-    spec fun is_currency {
-        ensures result == spec_is_currency<CoinType>();
+    spec module {
+        /// The invariant does not hold right now because an account doesn't
+        /// have to be the treasury compliance account to have MintCapability.
+        /// In this module, any account with PrivilegedCapability<Association>
+        /// can be granted the MintCapability, and in Association.move, any
+        /// account can be granted PrivilegedCapability<Association> through
+        /// `grant_privilege`. Therefore, we cannot prove that only the
+        /// treasury compliance account has MintCapability.
+        apply OnlyTCHasMintCapabilityInvariant to *, *<CoinType>;
     }
 
-    // Move code
-    spec fun assert_is_coin {
-        aborts_if !spec_is_currency<CoinType>();
+    spec schema OnlyTCHasBurnCapabilityInvariant {
+        /// Before a currency is registered, there is no burn capability for that currency.
+        invariant module forall coinType: type, addr: address:
+            !spec_is_currency<coinType>() ==> !exists<BurnCapability<coinType>>(addr);
+
+        /// After a currency is registered, only the treasury compliance account
+        /// has the mint capability for that currency.
+        invariant module forall coinType: type, addr: address
+        where spec_is_currency<coinType>():
+            exists<BurnCapability<coinType>>(addr)
+                ==> addr == spec_treasury_compliance_addr();
+    }
+
+    spec module {
+        /// Similar to OnlyTCHasMintCapabilityInvariant, this invariant
+        /// cannot be proved right now.
+        apply OnlyTCHasBurnCapabilityInvariant to *, *<CoinType>;
     }
 
-    // Maintain a ghost variable representing the sum of
-    // all coins of a currency type.
+    /// ## Conservation of currency
+
     spec module {
+        /// Maintain a ghost variable representing the sum of
```

Comment:
Lets not call this `ghost` variable. It is a confusing notion. We call it in the docs `spec variable`.